<a href="https://colab.research.google.com/github/HEOJIBEOM/Maritime-Data-Minin/blob/main/4%EC%A3%BC%EC%B0%A8%20%EA%B3%BC%EC%A0%9C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [11]:
# 라이브러리 로드
import numpy as np
import pandas as pd
import statsmodels.api as sm
from sklearn.model_selection import train_test_split   #sklearn.model_selection library 의 train_test_split() 함수를 사용하는 것을 알림.
from sklearn.metrics import mean_squared_error

pandas는 데이터 분석 및 처리를 위한 라이브러리로 데이터프레임을 다룰 때 사용한다.

as는 약칭을 지정하는 의미이다.

statsmodels.api는 통계 모델 수학적 알고리즘을 사용하여 모델링을 위한 라이브러리이다.

numpy는 수치 연산 및 다차원 배열을 처리하기 위한 라이브러리이다.

train_test_split은 데이터셋을 학습 및 테스트 세트로 분리하기 위한 함수이다.

mean_squared_error는 회귀 모델의 평가 지표로 사용되는 MSE(평균제곱오차)를 계산하는 함수이다.

In [4]:
#데이터 로드 및 확인
data = pd.read_csv("/content/02.야구선수연봉(투수).csv")
print(data.shape)
print(data.info())
print(data.head())

(152, 22)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 152 entries, 0 to 151
Data columns (total 22 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PLAYER       152 non-null    object 
 1   TEAM         152 non-null    object 
 2   VICTORY      152 non-null    int64  
 3   DEFEAT       152 non-null    int64  
 4   SAVE         152 non-null    int64  
 5   HOLD         152 non-null    int64  
 6   BLOWN        152 non-null    int64  
 7   GAME         152 non-null    int64  
 8   STARTER      152 non-null    int64  
 9   INNING       152 non-null    float64
 10  STRIKEOUT9   152 non-null    float64
 11  BALLNET9     152 non-null    float64
 12  HOMRUN9      152 non-null    float64
 13  BABIP        152 non-null    float64
 14  LOB          152 non-null    float64
 15  ERA          152 non-null    float64
 16  RA9_WAR      152 non-null    float64
 17  FIP          152 non-null    float64
 18  kFIP         152 non-null    float64
 19

read_csv() 함수를 이용하여 CSV 파일을 데이터프레임으로 불러와 data라는 변수에 저장한다.

data.shape는 데이터의 행과 열의 수를 출력하며, data.info()로 데이터의 기본 정보(컬럼명, 데이터 타입, null 값 유무 등)를 출력한다.

data.head()는 데이터프레임의 처음부터 일부 데이터만을 출력하는 데 사용되며, 디폴트는 5이다.

data.info(): DataFrame의 각 컬럼별 정보 (이름, Non-Null Count, 데이터 타입)를 확인하여 데이터의 기본적인 속성과 결측치 여부, 데이터 타입을 파악합니다.

"/content/02.야구선수연봉(투수).csv": read_csv() 함수의 인자로 전달된 파일 경로입니다. 이 경로는 Google Colab 환경의 /content/ 폴더에 있는 02.야구선수연봉(투수).csv 파일을 읽어오라는 의미입니다.

In [5]:
#train/test 데이터셋 분리
train_data, test_data = train_test_split(data, test_size=0.2, random_state=1)

train_data: 전체 데이터의 80%를 무작위로 추출하여 구성된 학습용 데이터셋입니다. 이 데이터는 모델을 학습시키는 데 사용됩니다.

test_data: 전체 데이터의 20%를 무작위로 추출하여 구성된 테스트용 데이터셋입니다. 이 데이터는 학습된 모델의 성능을 평가하는 데 사용됩니다. 학습 과정에 직접적으로 사용되지 않고, 모델이 처음 보는 새로운 데이터에 대해 얼마나 잘 예측하는지를 측정하는 데 활용됩니다.

random_state: 데이터셋을 무작위로 분리할 때 사용되는 난수 생성기의 시드(seed) 값을 지정하는 파라미터입니다.

In [6]:
#회귀모델 생성
#모델1
model1 = sm.OLS.from_formula("SALARY_2018 ~ " + " + ".join(train_data.columns.difference(['SALARY_2018'])), data=train_data).fit()
print(model1.summary())

#모델2
model2 = sm.OLS.from_formula("SALARY_2018 ~ VICTORY + DEFEAT + SAVE + WAR", data=train_data).fit()
print(model2.summary())


                            OLS Regression Results                            
Dep. Variable:            SALARY_2018   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Mon, 31 Mar 2025   Prob (F-statistic):                nan
Time:                        08:09:51   Log-Likelihood:                 2057.6
No. Observations:                 121   AIC:                            -3873.
Df Residuals:                       0   BIC:                            -3535.
Df Model:                         120                                         
Covariance Type:            nonrobust                                         
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
Intercept           

/usr/local/lib/python3.11/dist-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/usr/local/lib/python3.11/dist-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/usr/local/lib/python3.11/dist-packages/statsmodels/regression/linear_model.py:1717: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid


모델1

sm.OLS.from_formula(): statsmodels.api에서 제공하는 Ordinary Least Squares (OLS, 최소자승법) 회귀 모델을 생성하는 함수입니다. 이 함수는 종속 변수와 독립 변수 간의 관계를 formula (수식) 형태로 지정하는 방식을 사용합니다.

"SALARY_2018 ~ " + " + ".join(train_data.columns.difference(['SALARY_2018'])): 이는 회귀 모델의 formula를 정의하는 부분입니다.

"SALARY_2018 ~ ": SALARY_2018 컬럼을 종속 변수(예측하고자 하는 값)로 설정한다는 의미입니다. ~ 기호는 종속 변수와 독립 변수를 구분합니다.

train_data.columns.difference(['SALARY_2018']):
train_data.columns: 학습 데이터셋(train_data)의 모든 컬럼 이름들을 담고 있는 pandas Index 객체를 반환합니다.

.difference(['SALARY_2018']): 이 Index 객체에서 ['SALARY_2018'] 리스트에 있는 컬럼 이름(즉, 종속 변수 이름)을 제외한 나머지 컬럼 이름들을 담고 있는 새로운 Index 객체를 반환합니다. 이는 종속 변수를 제외한 모든 컬럼을 독립 변수로 사용하겠다는 의미입니다.

" + ".join(...): 위에서 얻은 독립 변수들의 컬럼 이름들을 + 기호로 연결하여 하나의 문자열로 만듭니다. 이는 회귀 모델의 formula에서 여러 독립 변수들을 + 기호로 연결하여 나타내는 방식입니다.

전체 formula: 결과적으로 모델 1은 SALARY_2018을 종속 변수로 하고, 학습 데이터셋(train_data)의 SALARY_2018 컬럼을 제외한 모든 다른 컬럼들을 독립 변수로 사용하는 선형 회귀 모델을 정의합니다.

, data=train_data: 생성할 회귀 모델에 사용할 데이터로 학습 데이터셋(train_data)을 지정합니다.

.fit(): 정의된 OLS 회귀 모델을 학습 데이터(train_data)를 사용하여 적합(fitting)시킵니다. 즉, 최소자승법을 이용하여 회귀 계수들을 추정합니다. 학습된 모델 객체가 model1 변수에 할당됩니다.

print(model1.summary()): 학습된 회귀 모델 model1의 요약 결과를 출력합니다. summary() 메서드는 모델의 계수 추정치, 표준 오차, t-통계량, p-값, 결정 계수(R-squared), 수정된 결정 계수(Adjusted R-squared), F-통계량 등 회귀 분석 결과를 상세하게 보여줍니다. 이를 통해 모델의 통계적 유의성과 설명력을 평가할 수 있습니다.

모델2

sm.OLS.from_formula("SALARY_2018 ~ VICTORY + DEFEAT + SAVE + WAR", data=train_data).fit(): 모델 1과 마찬가지로 OLS 회귀 모델을 생성하고 학습시키는 과정입니다. 하지만 formula가 다릅니다.

"SALARY_2018 ~ VICTORY + DEFEAT + SAVE + WAR": 이번에는 종속 변수는 여전히 SALARY_2018이지만, 독립 변수로는 VICTORY, DEFEAT, SAVE, WAR 네 개의 특정 컬럼만을 사용하도록 명시적으로 지정했습니다.
, data=train_data: 학습 데이터셋(train_data)을 모델 학습에 사용합니다.

.fit(): 정의된 회귀 모델을 학습 데이터로 적합시킵니다. 학습된 모델 객체가 model2 변수에 할당됩니다.

print(model2.summary()): 학습된 회귀 모델 model2의 요약 결과를 출력합니다. 모델 1과 마찬가지로 계수 추정치, 통계적 유의성 지표, 모델 설명력 등을 확인할 수 있습니다.

정리

두 모델의 summary() 결과를 비교 분석함으로써 어떤 변수들이 연봉 예측에 유의미한 영향을 미치는지, 모델의 설명력은 어느 정도인지 등을 파악하고, 더 나아가 더 나은 예측 모델을 구축하기 위한 방향을 설정할 수 있습니다. 예를 들어, 모델 1의 결과를 보고 유의미하지 않은 변수들을 제거하거나, 모델 2에 다른 중요한 변수들을 추가하는 등의 개선 작업을 수행할 수 있습니다.

In [7]:
# 예측
pre_lm = model2.predict(test_data)
result = pd.DataFrame({'actual': test_data['SALARY_2018'], 'predict': pre_lm})
print(result)


     actual       predict
59    10000  39974.619819
33    20000  14498.889582
69     8500  11656.423474
127    3200   3660.275219
31    20000  49535.237615
107    4000  11551.607630
97     5200  20753.344459
14    50000  34522.360453
53    11500  33330.787315
44    15000  27195.257084
146    2800   5979.879921
138    3000   2355.497574
78     7100  -5343.083468
102    4500   2232.154340
5    100000  87538.406815
85     6300   4049.414877
16    50000   7699.066206
100    4700  -4954.818843
89     6000   6993.202690
95     5400   6848.227396
91     6000    123.828397
124    3300   4675.102276
92     5800   8883.508283
99     4700  16051.468513
93     5700  -2449.204364
40    15500  52790.122512
29    23000  23944.153710
42    15000   8103.096332
54    11000  39639.359853
122    3400   5840.531400
35    19000  20492.785338


model2: 앞선 코드에서 sm.OLS.from_formula()를 사용하여 학습시킨 선형 회귀 모델 객체입니다. 이 모델은 VICTORY, DEFEAT, SAVE, WAR를 독립 변수로 사용하여 SALARY_2018을 예측하도록 학습되었습니다.

.predict(test_data): 학습된 model2 모델의 predict() 메서드를 호출합니다. 이 메서드는 새로운 데이터(여기서는 test_data)를 입력받아 종속 변수(SALARY_2018)의 예측값을 생성합니다.

test_data: 학습 과정에 사용되지 않았던 테스트 데이터셋입니다. 모델의 일반화 성능을 평가하기 위해 사용됩니다. predict() 메서드는 test_data의 독립 변수(VICTORY, DEFEAT, SAVE, WAR) 컬럼의 값을 사용하여 각 선수에 대한 2018년 연봉을 예측합니다.

pre_lm: predict() 메서드의 결과로 반환되는 예측된 연봉 값들이 pre_lm이라는 변수에 저장됩니다. 이는 NumPy 배열 또는 pandas Series 형태일 가능성이 높습니다.

pd.DataFrame({...}): pandas의 DataFrame 객체를 생성합니다. DataFrame은 테이블 형태의 데이터를 다루는 데 유용한 자료 구조입니다.

{'actual': test_data['SALARY_2018'], 'predict': pre_lm}: DataFrame의 내용을 정의하는 딕셔너리입니다.

'actual': test_data['SALARY_2018']: 딕셔너리의 'actual'이라는 키에 테스트 데이터셋(test_data)의 실제 SALARY_2018 컬럼의 값들을 할당합니다. 이는 모델이 예측해야 했던 정답(ground truth)에 해당합니다.

'predict': pre_lm: 딕셔너리의 'predict'이라는 키에 앞서 모델이 예측한 연봉 값들이 저장된 pre_lm 변수를 할당합니다.

result: 생성된 DataFrame 객체가 result라는 변수에 할당됩니다. 이 DataFrame은 'actual' 컬럼에 실제 연봉, 'predict' 컬럼에 모델이 예측한 연봉을 나란히 보여줍니다.

정리

개별 예측 결과 확인: 각 테스트 데이터에 대해 모델이 얼마나 정확하게 연봉을 예측했는지 'actual' 컬럼과 'predict' 컬럼을 비교하여 알 수 있습니다.
예측 오차의 방향 및 크기 파악: 실제 연봉보다 높게 예측했는지 낮게 예측했는지, 그리고 그 차이가 얼마나 되는지를 확인할 수 있습니다.
모델의 전반적인 경향성 파악: 전체적인 예측 결과의 패턴을 보면서 모델이 실제 연봉을 과대 또는 과소 예측하는 경향이 있는지 등을 대략적으로 파악할 수 있습니다.


In [8]:
# RMSE 계산
import math
rmse_value = math.sqrt(mean_squared_error(test_data['SALARY_2018'], pre_lm))
print("RMSE:", rmse_value)


RMSE: 15760.477097309951


import math: 파이썬의 표준 수학 라이브러리를 불러옵니다. 여기서는 제곱근을 계산하는 math.sqrt() 함수를 사용하기 위해 임포트되었습니다.

mean_squared_error(): 앞서 코드의 첫 부분에서 from sklearn.metrics import mean_squared_error를 통해 불러온 함수입니다.

test_data['SALARY_2018']: 테스트 데이터셋의 실제 2018년 연봉 값들을 나타냅니다.

pre_lm: 앞선 코드에서 model2.predict(test_data)를 통해 얻은 예측된 2018년 연봉 값들을 나타냅니다.
이 함수는 실제 연봉과 예측 연봉 간의 평균 제곱 오차 (MSE)를 계산하여 반환합니다. 즉, 각 데이터 포인트에 대해 (실제 값 - 예측 값)² 을 계산하고, 이 값들의 평균을 구합니다.

math.sqrt(): math 라이브러리에서 제공하는 제곱근을 계산하는 함수입니다.
mean_squared_error() 함수의 반환값인 MSE를 입력으로 받아 그 제곱근을 계산합니다. 이 값이 바로 RMSE입니다.
rmse_value = ...: 계산된 RMSE 값이 rmse_value라는 변수에 할당됩니다.

정리

RMSE 값 15760.48은 모델 2가 VICTORY, DEFEAT, SAVE, WAR 변수만을 사용하여 2018년 연봉을 예측했을 때, 실제 연봉과 예측 연봉 간의 평균적인 오차가 약 15760.48 (데이터 단위에 따라 해석) 만큼 발생한다는 것을 의미합니다. 이 값이 모델의 성능이 좋다고 단정하기는 어렵고, 데이터의 특성, 다른 모델과의 비교, 비즈니스적인 중요도 등을 종합적으로 고려하여 판단해야 합니다.